---
title: "WGS-SNP summary"
date: "`r Sys.Date()`"
output: html_document
params:
  results_logs: ""
  results_vcfs: ""
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("data.table")
library("DT")
library("ggplot2")
library("plotly")
library("reshape2")
```

```{r import_data, echo=FALSE}
# read data
df1 <- fread(params$results_logs)
mdf <- reshape2::melt(df1, id = "id")

df2 <- fread(params$results_vcfs)
```
# {.tabset}
## BWA-MEM alignment results

Plot showing the amount of total and mapped reads per sample.

```{r alignment_logs, echo=FALSE}
p <- ggplot(mdf,aes(x=id,y=value,fill=variable)) +
  geom_bar(stat="identity",position = "identity") +
  scale_fill_manual(values = c("aquamarine3", "cornflowerblue")) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -40, vjust = 1, hjust = 0)) + 
  xlab("Sample") + 
  ylab("# reads") +
  labs(fill=NULL)
  
ggplotly(p)
```

Table of reads

```{r alignment_table, echo=FALSE}
colnames(df1) <- c("Sample", "Total reads", "Mapped reads")
datatable(df1)
```
## SNP analysis

Table of SNP variants across all samples.

TRUE or FALSE indicate presence or absence of the alternate allele.

```{r vcf_analysis, echo=FALSE}
datatable(df2,
          extensions = "Buttons",
          options = list(
            dom = "Bfrtlip",
            buttons = c('csv', 'excel'),
            pageLength = 25,
            columnDefs = list(list(
               targets = -1,
               render = JS(
                 "function(data, type, row, meta) {",
                 "return type === 'display' && data.length > 150 ?",
                 "'<span title=\"' + data + '\">' + data.substr(0, 150) + '...</span>' : data;",
                 "}")
             )))
          )
```

CHROM: Chromosome or plasmid

POS: Position in the chromosome or plasmid

REF: Reference allele

ALT: Alternate allele

ANN: Functional annotations (if SnpEff option was selected/available): 'Allele | Annotation | Annotation_Impact | Gene_Name | Gene_ID | Feature_Type | Feature_ID | Transcript_BioType | Rank | HGVS.c | HGVS.p | cDNA.pos / cDNA.length | CDS.pos / CDS.length | AA.pos / AA.length | Distance | ERRORS / WARNINGS / INFO'

## Methods

Nextflow pipeline for whole genome sequencing (WGS) and variant calling of bacterial genomes.

Tools and references:
- Nextflow version 24.04.4.5917
- Docker version 27.2.0
- NCBI genome assemblies (https://www.ncbi.nlm.nih.gov/datasets/genome/)
- BWA-MEM version 0.7.18 (https://bio-bwa.sourceforge.net/)
- samtools version 1.20:
  - samtools/flagstat (http://www.htslib.org/doc/samtools-flagstat.html)
  - samtools/sort (http://www.htslib.org/doc/samtools-sort.html)
- Picard/MarkDuplicates version 3.1.1 (https://broadinstitute.github.io/picard/)
- Freebayes version 1.3.6 (https://github.com/freebayes/freebayes)
- VCFtools version 0.1.16 (https://vcftools.github.io/index.html)
- VCF-SnpEff formatter version 0.4.0 (https://github.com/lalvarezmunoz/vcf-snpeff-formatter)
- SnpEff version 5.2 (https://pcingola.github.io/SnpEff/snpeff/introduction/)
- R version 4.4.2 (with packages: remotes, data.table v1.16.4, dplyr v1.1.4, DT v0.33, ggplot2 v3.5.1, knitr v1.49, plotly v4.10.4, reshape2 v1.4.4, rmarkdown v2.29, vcfR v1.15.0) 

### Citing this pipeline

Please, refer to the GitHub repository when using this pipeline:

> L. Alvarez, nf-wgs-snp pipeline, (2024), GitHub repository, https://github.com/lalvarezmunoz/nf-wgs-snp